src/lib.rs

```diff
@@ -0,0 +1,902 @@
+    }
+
+    #[inline(always)]
+    pub fn kind(self) -> AllocationKind {
+        match (self.bits >> 29) & 7 {
+            5 => AllocationKind::None,
+            6 => AllocationKind::Reg,
+            7 => AllocationKind::Stack,
+            _ => unreachable!(),
+        }
+    }
+
+    #[inline(always)]
+    pub fn is_none(self) -> bool {
+        self.kind() == AllocationKind::None
+    }
+
+    #[inline(always)]
+    pub fn is_reg(self) -> bool {
+        self.kind() == AllocationKind::Reg
+    }
+
+    #[inline(always)]
+    pub fn is_stack(self) -> bool {
+        self.kind() == AllocationKind::Stack
+    }
+
+    #[inline(always)]
+    pub fn index(self) -> usize {
+        (self.bits & ((1 << 28) - 1)) as usize
+    }
+
+    #[inline(always)]
+    pub fn as_reg(self) -> Option<PReg> {
+        if self.kind() == AllocationKind::Reg {
+            Some(PReg::from_index(self.index()))
+        } else {
+            None
+        }
+    }
+
+    #[inline(always)]
+    pub fn as_stack(self) -> Option<SpillSlot> {
+        if self.kind() == AllocationKind::Stack {
+            Some(SpillSlot(self.index() as u32))
+        } else {
+            None
+        }
+    }
+
+    #[inline(always)]
+    pub fn bits(self) -> u32 {
+        self.bits
+    }
+
+    #[inline(always)]
+    pub fn from_bits(bits: u32) -> Self {
+        Self { bits }
+    }
+}
+
+// N.B.: These values must be *disjoint* with the values used to
+// encode `OperandPolicy`, because they share a 3-bit field.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[repr(u8)]
+pub enum AllocationKind {
+    None = 5,
+    Reg = 6,
+    Stack = 7,
+}
+
+impl Allocation {
+    #[inline(always)]
+    pub fn class(self) -> RegClass {
+        match self.kind() {
+            AllocationKind::None => panic!("Allocation::None has no class"),
+            AllocationKind::Reg => self.as_reg().unwrap().class(),
+            AllocationKind::Stack => self.as_stack().unwrap().class(),
+        }
+    }
+}
+
+/// A helper that wraps either an `Operand` or an `Allocation` and is
+/// able to tell which it is based on the tag bits.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct OperandOrAllocation {
+    bits: u32,
+}
+
+impl OperandOrAllocation {
+    pub fn from_operand(operand: Operand) -> Self {
+        Self {
+            bits: operand.bits(),
+        }
+    }
+    pub fn from_alloc(alloc: Allocation) -> Self {
+        Self { bits: alloc.bits() }
+    }
+    pub fn is_operand(&self) -> bool {
+        (self.bits >> 29) <= 4
+    }
+    pub fn is_allocation(&self) -> bool {
+        (self.bits >> 29) >= 5
+    }
+    pub fn as_operand(&self) -> Option<Operand> {
+        if self.is_operand() {
+            Some(Operand::from_bits(self.bits))
+        } else {
+            None
+        }
+    }
+    pub fn as_allocation(&self) -> Option<Allocation> {
+        if self.is_allocation() {
+            Some(Allocation::from_bits(self.bits & !(1 << 28)))
```

Comment:
Hmm, yeah, I clarified the comment in `Allocation` and added one here: the intent is actually for this bit to belong to the `OperandOrAllocation`, i.e., the wrapper knows whether the `Allocation` is a def or use, but the `Allocation` itself does not have that bit. I want the latter to be true because allocs are just names for locations. ("User-defined" was from when I had put this wrapper on the Cranelift side, but it seems to fit better here.)